aten/src/ATen/native/Linear.cpp

```diff
@@ -136,241 +136,331 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
+        nlabels,
+        has_ellipsis ? ") is more than the number of dimensions ("
+                     : ") does not match the number of dimensions (",
+        ndims,
+        ") for operand ",
+        i,
+        has_ellipsis ? "" : " and no ellipsis was given");
+  }
+
+  // Mapping of label to index in the permuted tensors (out_dims + sum_dims)
+  // This will be used for aligning the dimensions of all input operands
+  std::vector<int> label_perm_index(total_labels, -1);
+  
+  // Current index in the permuted shape
+  int perm_index = 0;
+
+  // Start index of ellipsis dimensions in the permuted shape
+  int64_t ell_index = 0;
+
+  if (arrow_pos == std::string::npos) {
+    // Implicit output is ellipsis (...) + labels seen only once
+    perm_index = ell_num_dim;
+    for (int label = 0; label < total_labels; ++label) {
+      if (label_count[label] == 1) {
+        label_perm_index[label] = perm_index++;
       }
     }
-    // then the indices that occur exactly once in alphabetic order
-    for (size_t idx = 0; idx < number_of_letters; idx++) {
-      if (num_letter_occurrences[idx] == 1) {
-        idxes_to_preprocessed_dims[letter_mapping[idx]] = num_output_dims;
-        num_output_dims++;
+  } else {
+    // Parse explicit output
+    std::string rhs = equation.substr(arrow_pos + 2);
+    found_ell = false;
+    for (std::size_t i = 0; i < rhs.length(); ++i) {
+      switch (rhs[i]) {
+        case ' ':
+          // Ignore spaces
+          break;
+
+        case '.':
+          TORCH_CHECK(
+              // There can only be one ellipsis in the output
+              !found_ell,
+              "einsum() found \'.\' for output but an ellipsis (...) was already found");
+          TORCH_CHECK(
+              // Ensure ellipsis is correct
+              i + 2 < rhs.length() && rhs[++i] == '.' && rhs[++i] == '.',
+              "einsum() found \'.\' for output that is not part of any ellipsis (...)");
+          ell_index = perm_index;
+          perm_index += ell_num_dim;
+          found_ell = true;
+          break;
+
+        default:
+          TORCH_CHECK(
+              rhs[i] >= 'a' && rhs[i] <= 'z',
+              "einsum() subscripts must be in range [a, z] but found ",
+              rhs[i],
+              " for the output");
+          TORCH_CHECK(
+              // Ensure label appeared at least once for some input operand and at
+              // most once for the output
+              label_count[rhs[i] - 'a'] > 0,
+              "einsum() output subscript ",
+              rhs[i],
+              label_count[rhs[i] - 'a'] == -1
+                  ? " appears more than once in the output"
+                  : " does not appear in the equation for any input operand");
+          label_perm_index[rhs[i] - 'a'] = perm_index++;
+          
+          // Set to -1 to mark that this label already appeared in the output
+          label_count[rhs[i] - 'a'] = -1;
       }
     }
+
+    TORCH_CHECK(
```

Comment:
I think this check is breaking Opacus

Reply:
Hey @Darktex, could you elaborate?

Reply:
It could be, I didn't notice this at first but the previous version does not enforce this rule so this check is BC breaking. @mruberry I can send a quick patch to remove this rule and allow ellipsis to be reduced or I can add a BC breaking note. (numpy has this enforcement but it doesn't seem necessary).